code/game/atoms_movable.dm

```diff
@@ -375,20 +376,35 @@
 	loc = new_loc
 	Moved(old_loc)
 
-////////////////////////////////////////
-// Here's where we rewrite how byond handles movement except slightly different
-// To be removed on step_ conversion
-// All this work to prevent a second bump
-/atom/movable/Move(atom/newloc, direction, glide_size_override = 0)
-	. = FALSE
-	if(!newloc || newloc == loc)
-		return
+#define SET_CARDINAL_DIR(set_dir_on_move, direction, can_pass_diagonally)\
+if(set_dir_on_move){setDir(direction &~ can_pass_diagonally)}
+
+/atom/movable/Move(atom/newloc, direction, glide_size_override)
+	var/atom/movable/pullee = pulling
+	if(!moving_from_pull)
+		check_pulling()
+	if(!loc || !newloc || loc == newloc)
+		return FALSE
 
 	if(!direction)
 		direction = get_dir(src, newloc)
 
-	if(set_dir_on_move)
-		setDir(direction)
+	var/can_pass_diagonally = NONE
+	if (direction & (direction - 1)) //Check if the first part of the diagonal move is possible
+		moving_diagonally = TRUE
+		if(set_dir_on_move)
+			setDir(direction) //We first set the direction to prevent going through dir sensible object
+		if(direction & NORTH)
+			can_pass_diagonally = get_step(loc, NORTH)?.Enter(src) ? NORTH : NONE
+		if(!can_pass_diagonally && (direction & EAST))
+			can_pass_diagonally = get_step(loc, EAST)?.Enter(src) ? EAST : NONE
+		if(!can_pass_diagonally && (direction & WEST))
+			can_pass_diagonally = get_step(loc, WEST)?.Enter(src) ? WEST : NONE
+		if(!can_pass_diagonally && (direction & SOUTH))
+			can_pass_diagonally = get_step(loc, SOUTH)?.Enter(src) ? SOUTH : NONE
+		moving_diagonally = FALSE
+		if(!can_pass_diagonally)
+			return
```

Comment:
```suggestion
		if((direction & NORTH) && get_step(loc, NORTH)?.Enter(src))
			can_pass_diagonally = NORTH
		else if((direction & EAST) && get_step(loc, EAST)?.Enter(src))
			can_pass_diagonally =  EAST
		else if((direction & WEST) && get_step(loc, WEST)?.Enter(src))
			can_pass_diagonally = WEST
		else if((direction & SOUTH) && get_step(loc, SOUTH)?.Enter(src))
			can_pass_diagonally = SOUTH
		else
			moving_diagonally = FALSE
			return
		moving_diagonally = FALSE
```
Made this code a tiny bit more succinct.

Reply:
yeah, much better. Thanks